src/Analysis/Ast/Impl/Generators/StubGenerator.CleanupWhitespaceWalker.cs

```diff
@@ -0,0 +1,226 @@
+                    switch (node) {
+                        case PythonAst _:
+                            return false;
+                        case SuiteStatement _:
+                            return false;
+                        case Statement _:
+                            return true;
+                        default:
+                            return false;
+                    }
+                }
+            }
+
+            private static IndexSpan GetSpan(Node previous, Node current) {
+                if (previous == null) {
+                    return IndexSpan.FromBounds(0, current.StartIndex);
+                }
+
+                // previous could contain current (ex, class definition)
+                // in that case, break span from previous.Start to current.start
+                // rather than end
+                if (previous.IndexSpan.Contains(current.IndexSpan)) {
+                    return IndexSpan.FromBounds(GetEndIndex(previous), current.StartIndex);
+                }
+
+                return IndexSpan.FromBounds(previous.EndIndex, current.StartIndex);
+
+                int GetEndIndex(Node node) {
+                    if (node is ClassDefinition @class) {
+                        return @class.HeaderIndex + 1;
+                    }
+
+                    if (node is FunctionDefinition func) {
+                        return func.HeaderIndex + 1;
+                    }
+
+                    return node.EndIndex;
+                }
+            }
+
+            private void PushIndentation(ScopeStatement node, int headerIndex) {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                _indentationStack.Push(ComputeIndentation(node));
+                _indentations.AddLast((headerIndex, _indentationStack.Peek()));
+            }
+
+            private void PopIndentation(ScopeStatement node, Node parent) {
+                CancellationToken.ThrowIfCancellationRequested();
+
+                _indentationStack.Pop();
+
+                var indentation = _indentationStack.Count == 0 ? 0 : _indentationStack.Peek();
+                _indentations.AddLast((node.EndIndex, indentation));
+
+                Debug.Assert(indentation == ComputeIndentation(GetContainer(parent)));
+            }
+
+            private string GetSpacesBetween(Node previous, Node current, int indentation) {
+                if (previous == null) {
+                    return new string(' ', indentation);
+                }
+
+                // same kind of node of one liner
+                if (previous.NodeName == current.NodeName && OnSingleLine(previous) && OnSingleLine(current)) {
+                    return Environment.NewLine + new string(' ', indentation);
+                }
+
+                // no line between header and doc comment
+                if (IsDocumentation(current as Statement)) {
+                    return Environment.NewLine + new string(' ', indentation);
```

Comment:
I think `$"{Environment.NewLine}{...}"` may be preferable (and few other string + string), but up to you.